torch/csrc/jit/codegen/cuda/index_compute.cpp

```diff
@@ -1030,6 +1095,14 @@ std::vector<kir::Val*> Index::getGlobalProducerStridedIndices(
 
     root_ind = getProducerIndexWithHalo(producer_tv, i, root_ind, consumer_tv);
 
+    root_ind = getProducerIndexWithGather(
```

Comment:
Yes, and that's actually a common case. When an intermediate tensor is gathered, it's expanded with halo to accommodate the neighbor accesses, so the gather producer does have halo, and `getProducerIndexWithHalo` should adjust the producer indexing for the hallo attached to the producer. And since the expression itself is `gather`, `getProducerIndexWithGather` should adjust the index accordingly. 

These parts of indexing are getting disorganized as I add more operations. For producer indexing, the adjustments we need to do are done when:

- when the producer has halo
- when the expression is shift
- when the expression is gather

Previously, both of the halo and shift adjustment are done in `getProducerHaloOffset`. For more clarity, the shift part of the adjustment may be better if it's extracted out of the function.